File: reminder/internal/telemetry/telemetry.go
// Package telemetry provides OpenTelemetry instrumentation for the worker service.
package telemetry

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	otelpkg "go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"

	appconfig "github.com/docvault/reminder/internal/config"
)

// Provider holds OpenTelemetry providers and instruments.
type Provider struct {
	TracerProvider *sdktrace.TracerProvider
	MeterProvider  *sdkmetric.MeterProvider
	Tracer         trace.Tracer
	Meter          otelpkg.Meter

	// Metrics
	JobCount        otelpkg.Int64Counter
	JobDuration     otelpkg.Float64Histogram
	JobErrorCount   otelpkg.Int64Counter
	ActiveJobs      otelpkg.Int64UpDownCounter
	DLQMessageCount otelpkg.Int64Counter
}

var globalProvider *Provider

// Init initializes OpenTelemetry with OTLP exporters.
func Init(ctx context.Context, cfg *appconfig.Config, serviceName string) (*Provider, error) {
	endpoint := cfg.OTELEndpoint
	if endpoint == "" {
		slog.Warn("OTEL_ENDPOINT not configured, skipping telemetry initialization")
		return &Provider{}, nil
	}

	// Create OTLP trace exporter
	traceExporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(endpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create trace exporter: %w", err)
	}

	// Create OTLP metric exporter
	metricExporter, err := otlpmetricgrpc.New(ctx,
		otlpmetricgrpc.WithEndpoint(endpoint),
		otlpmetricgrpc.WithInsecure(),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create metric exporter: %w", err)
	}

	// Create resource with service information
	res, err := resource.New(ctx,
		resource.WithAttributes(
			attribute.String("service.name", serviceName),
			attribute.String("service.version", "1.0.0"),
			attribute.String("environment", cfg.Environment),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create resource: %w", err)
	}

	// Create tracer provider
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)

	// Create meter provider
	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExporter)),
	)

	otel.SetTracerProvider(tp)
	otel.SetMeterProvider(mp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	// Create tracer and meter
	tracer := tp.Tracer(serviceName)
	meter := mp.Meter(serviceName)

	// Create metrics instruments
	jobCount, err := meter.Int64Counter(
		"worker_jobs_total",
		otelpkg.WithDescription("Total number of jobs processed"),
		otelpkg.WithUnit("{job}"),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create job count metric: %w", err)
	}

	jobDuration, err := meter.Float64Histogram(
		"worker_job_duration_seconds",
		otelpkg.WithDescription("Job processing duration in seconds"),
		otelpkg.WithUnit("s"),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create job duration metric: %w", err)
	}

	jobErrorCount, err := meter.Int64Counter(
		"worker_jobs_errors_total",
		otelpkg.WithDescription("Total number of job errors"),
		otelpkg.WithUnit("{error}"),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create job error metric: %w", err)
	}

	activeJobs, err := meter.Int64UpDownCounter(
		"worker_active_jobs",
		otelpkg.WithDescription("Number of active jobs being processed"),
		otelpkg.WithUnit("{job}"),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create active jobs metric: %w", err)
	}

	dlqMessageCount, err := meter.Int64Counter(
		"worker_dlq_messages_total",
		otelpkg.WithDescription("Total number of messages sent to DLQ"),
		otelpkg.WithUnit("{message}"),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create DLQ metric: %w", err)
	}

	globalProvider = &Provider{
		TracerProvider:  tp,
		MeterProvider:   mp,
		Tracer:          tracer,
		Meter:           meter,
		JobCount:        jobCount,
		JobDuration:     jobDuration,
		JobErrorCount:   jobErrorCount,
		ActiveJobs:      activeJobs,
		DLQMessageCount: dlqMessageCount,
	}

	slog.Info("OpenTelemetry initialized",
		"service", serviceName,
		"endpoint", endpoint,
		"environment", cfg.Environment,
	)

	return globalProvider, nil
}

// Shutdown gracefully shuts down the telemetry providers.
func Shutdown(ctx context.Context) error {
	if globalProvider == nil {
		return nil
	}

	if globalProvider.TracerProvider != nil {
		if err := globalProvider.TracerProvider.Shutdown(ctx); err != nil {
			return fmt.Errorf("failed to shutdown tracer provider: %w", err)
		}
	}

	slog.Info("OpenTelemetry shutdown complete")
	return nil
}

// GetProvider returns the global telemetry provider.
func GetProvider() *Provider {
	return globalProvider
}

// StartSpan starts a new span with the given name.
func StartSpan(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
	if globalProvider == nil || globalProvider.Tracer == nil {
		return ctx, trace.SpanFromContext(ctx)
	}
	return globalProvider.Tracer.Start(ctx, name, opts...)
}

// RecordJob records a job processing metric.
func (p *Provider) RecordJob(ctx context.Context, jobType string, success bool, duration time.Duration) {
	if p == nil {
		return
	}

	attrs := []attribute.KeyValue{
		attribute.String("job.type", jobType),
		attribute.Bool("job.success", success),
	}

	p.JobCount.Add(ctx, 1, otelpkg.WithAttributes(attrs...))
	p.JobDuration.Record(ctx, duration.Seconds(), otelpkg.WithAttributes(attrs...))

	if !success {
		p.JobErrorCount.Add(ctx, 1, otelpkg.WithAttributes(attrs...))
	}
}

// RecordDLQ records a DLQ message metric.
func (p *Provider) RecordDLQ(ctx context.Context, jobType string) {
	attrs := []attribute.KeyValue{
		attribute.String("job.type", jobType),
	}
	p.DLQMessageCount.Add(ctx, 1, otelpkg.WithAttributes(attrs...))
}

// AddSpanEvent adds an event to the current span.
func AddSpanEvent(ctx context.Context, name string, attrs ...attribute.KeyValue) {
	span := trace.SpanFromContext(ctx)
	span.AddEvent(name, trace.WithAttributes(attrs...))
}

// SetSpanAttributes sets attributes on the current span.
func SetSpanAttributes(ctx context.Context, attrs ...attribute.KeyValue) {
	span := trace.SpanFromContext(ctx)
	span.SetAttributes(attrs...)
}

// RecordError records an error on the current span.
func RecordError(ctx context.Context, err error, attrs ...attribute.KeyValue) {
	span := trace.SpanFromContext(ctx)
	span.RecordError(err, trace.WithAttributes(attrs...))
}

File: shared/py/docvault_shared/telemetry.py
"""OpenTelemetry instrumentation for DocVault Python services."""

import logging
import os
import time
import structlog
from typing import Optional, Dict, Any
from functools import wraps

import sentry_sdk
from sentry_sdk.integrations.logging import LoggingIntegration

from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.semconv.resource import ResourceAttributes
from opentelemetry.propagate import set_global_textmap
from opentelemetry.propagators.b3 import B3MultiFormat
from opentelemetry.trace import Status, StatusCode

logger = structlog.get_logger(__name__)


class TelemetryProvider:
    """Holds OpenTelemetry providers and instruments."""

    def __init__(self):
        self.tracer_provider: Optional[TracerProvider] = None
        self.meter_provider: Optional[MeterProvider] = None
        self.tracer = None
        self.meter = None

        self.job_counter = None
        self.job_duration = None
        self.job_error_counter = None
        self.active_jobs = None

        self._initialized = False


_global_provider: Optional[TelemetryProvider] = None


def get_logger():
    """Return the shared structured logger for telemetry-aware handlers."""
    return logger


def init_telemetry(service_name: str = "docvault-service") -> TelemetryProvider:
    """Initialize OpenTelemetry with OTLP exporters."""
    global _global_provider

    otel_endpoint = os.getenv("OTEL_EXPORTER_ENDPOINT", "")

    if not otel_endpoint:
        logger.warning("OTEL_EXPORTER_ENDPOINT not configured, skipping telemetry initialization")
        _global_provider = TelemetryProvider()
        return _global_provider

    environment = os.getenv("ENVIRONMENT", "development")

    resource = Resource.create(
        {
            SERVICE_NAME: service_name,
            SERVICE_VERSION: "1.0.0",
            ResourceAttributes.DEPLOYMENT_ENVIRONMENT: environment,
        }
    )

    span_exporter = OTLPSpanExporter(
        endpoint=otel_endpoint,
        insecure=True,
    )

    metric_exporter = OTLPMetricExporter(
        endpoint=otel_endpoint,
        insecure=True,
    )

    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(BatchSpanProcessor(span_exporter))

    metric_reader = PeriodicExportingMetricReader(metric_exporter, export_interval_millis=60000)
    meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])

    trace.set_tracer_provider(tracer_provider)
    metrics.set_meter_provider(meter_provider)

    set_global_textmap(B3MultiFormat())

    tracer = trace.get_tracer(service_name)
    meter = metrics.get_meter(service_name)

    job_counter = meter.create_counter(
        name="jobs_total",
        description="Total number of jobs",
        unit="1",
    )

    job_duration = meter.create_histogram(
        name="job_duration_seconds",
        description="Job duration in seconds",
        unit="s",
    )

    job_error_counter = meter.create_counter(
        name="jobs_errors_total",
        description="Total number of job errors",
        unit="1",
    )

    active_jobs = meter.create_up_down_counter(
        name="active_jobs",
        description="Number of active jobs",
        unit="1",
    )

    _global_provider = TelemetryProvider()
    _global_provider.tracer_provider = tracer_provider
    _global_provider.meter_provider = meter_provider
    _global_provider.tracer = tracer
    _global_provider.meter = meter
    _global_provider.job_counter = job_counter
    _global_provider.job_duration = job_duration
    _global_provider.job_error_counter = job_error_counter
    _global_provider.active_jobs = active_jobs
    _global_provider._initialized = True

    logger.info(
        "OpenTelemetry initialized",
        service=service_name,
        endpoint=otel_endpoint,
        environment=environment,
    )

    init_sentry(service_name, environment)

    return _global_provider


def init_sentry(service_name: str, environment: str, sentry_dsn: str = "") -> None:
    """Initialize Sentry error tracking."""
    from .config import config as shared_config

    if not sentry_dsn:
        sentry_dsn = shared_config.sentry_dsn

    if not sentry_dsn:
        logger.warning("SENTRY_DSN not configured, skipping Sentry initialization")
        return

    sentry_logging = LoggingIntegration(
        level=logging.INFO,
        event_level=logging.ERROR,
    )

    sentry_sdk.init(
        dsn=sentry_dsn,
        environment=environment,
        release=f"{service_name}@1.0.0",
        integrations=[sentry_logging],
        traces_sample_rate=1.0,
        ignore_errors=[KeyboardInterrupt],
    )

    logger.info(
        "Sentry initialized",
        service=service_name,
        environment=environment,
    )


def shutdown_telemetry() -> None:
    """Gracefully shutdown the telemetry providers."""
    global _global_provider

    if _global_provider and _global_provider.tracer_provider:
        _global_provider.tracer_provider.shutdown()

    sentry_sdk.flush(timeout=5.0)

    logger.info("Telemetry shutdown complete")


def get_provider() -> TelemetryProvider:
    """Get the global telemetry provider."""
    global _global_provider
    return _global_provider


def start_span(
    name: str, kind=trace.SpanKind.INTERNAL, attributes: Optional[Dict[str, Any]] = None
):
    """Start a new span with the given name."""
    provider = get_provider()
    if provider and provider.tracer:
        return provider.tracer.start_as_current_span(
            name,
            kind=kind,
            attributes=attributes,
        )
    return trace.get_tracer("noop").start_as_current_span(name)


def record_job(
    job_type: str, success: bool, duration: float, attributes: Optional[Dict[str, Any]] = None
) -> None:
    """Record a job processing metric."""
    provider = get_provider()
    if not provider or not provider._initialized:
        return

    attrs = {
        "job.type": job_type,
        "job.success": success,
    }
    if attributes:
        attrs.update(attributes)

    provider.job_counter.add(1, attrs)
    provider.job_duration.record(duration, attrs)

    if not success:
        provider.job_error_counter.add(1, attrs)


def record_error(error: Exception, attributes: Optional[Dict[str, Any]] = None) -> None:
    """Record an error in the current span."""
    span = trace.get_current_span()
    if span and span.is_recording():
        span.set_status(Status(StatusCode.ERROR, str(error)))
        span.record_exception(error, attributes=attributes)

    with sentry_sdk.configure_scope() as scope:
        for key, value in (attributes or {}).items():
            scope.set_tag(key, str(value))
        sentry_sdk.capture_exception(error)


def set_user_context(user_id: str, tenant_id: str = None) -> None:
    """Set user context for Sentry error tracking."""
    if not user_id:
        return

    sentry_sdk.set_user({"id": user_id})
    if tenant_id:
        sentry_sdk.set_tag("tenant_id", tenant_id)


def add_tenant_context(tenant_id: str) -> None:
    """Add tenant context to Sentry scope."""
    if tenant_id:
        sentry_sdk.set_tag("tenant_id", tenant_id)


def traced(job_type: str):
    """Decorator to trace a function as a job."""

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            start_time = time.time()
            provider = get_provider()

            if provider and provider._initialized and provider.active_jobs:
                provider.active_jobs.add(1, {"job.type": job_type})

            with start_span(
                f"{job_type}_job", kind=trace.SpanKind.CONSUMER, attributes={"job.type": job_type}
            ):
                try:
                    result = func(*args, **kwargs)
                    duration = time.time() - start_time
                    record_job(job_type, True, duration)
                    return result
                except Exception as e:
                    duration = time.time() - start_time
                    record_job(job_type, False, duration)
                    record_error(e)
                    raise
                finally:
                    if provider and provider._initialized and provider.active_jobs:
                        provider.active_jobs.add(-1, {"job.type": job_type})

        return wrapper

    return decorator


class SpanContextLogger:
    """Context manager that adds trace context to logs."""

    def __init__(self, logger, span_name: str, attributes: Optional[Dict[str, Any]] = None):
        self.logger = logger
        self.span_name = span_name
        self.attributes = attributes or {}
        self.start_time = None

    def __enter__(self):
        self.start_time = time.time()
        span = trace.get_current_span()

        if span and span.is_recording():
            ctx = span.get_span_context()
            self.logger = self.logger.bind(
                trace_id=format(ctx.trace_id, "032x"),
                span_id=format(ctx.span_id, "016x"),
            )

        self.logger = self.logger.bind(**self.attributes)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        duration = time.time() - self.start_time

        if exc_type:
            self.logger.error(
                f"{self.span_name} failed",
                duration_ms=int(duration * 1000),
                error=str(exc_val),
            )
        else:
            self.logger.info(
                self.span_name,
                duration_ms=int(duration * 1000),
            )

        return False


def log_with_trace(logger, span_name: str, attributes: Optional[Dict[str, Any]] = None):
    """Create a logger context with trace correlation."""
    return SpanContextLogger(logger, span_name, attributes)
